warehouse/macaroons/services.py

```diff
@@ -104,40 +85,66 @@ def verify(self, raw_macaroon, context, principals, permission):
 
         Raises InvalidMacaroon if the macaroon is not valid.
         """
-        m = self._deserialize_raw_macaroon(raw_macaroon)
-        dm = self.find_macaroon(m.identifier.decode())
+        token = self._deserialize_raw_macaroon(raw_macaroon)
+        dm = self.find_macaroon(token.identifier)
 
         if dm is None:
             raise InvalidMacaroon("deleted or nonexistent macaroon")
 
-        verifier = Verifier(m, context, principals, permission)
-        if verifier.verify(dm.key):
-            dm.last_used = datetime.datetime.now()
-            return True
+        project = context.normalized_name
+
+        try:
+            token.check(key=dm.key, project=project)
+        except pypitoken.ValidationError as exc:
+            raise InvalidMacaroon(str(exc))
 
-        raise InvalidMacaroon("invalid macaroon")
+        dm.last_used = datetime.datetime.now()
+        return True
 
-    def create_macaroon(self, location, user_id, description, caveats):
+    def create_macaroon(self, domain, user_id, description, restrictions):
         """
         Returns a tuple of a new raw (serialized) macaroon and its DB model.
         The description provided is not embedded into the macaroon, only stored
         in the DB model.
         """
         user = self.db.query(User).filter(User.id == user_id).one()
 
-        dm = Macaroon(user=user, description=description, caveats=caveats)
+        identifier = uuid.uuid4()
+        key = _generate_key()
```

Comment:
The previous flow was

1. we generate the caveats object
2. we create the database macaroon, letting the DB choose the identifier and have the key computed by a python function set as model default
3. we read the id and key from the object after the DB flush
4. we generate the pymacaroons object from those

Because we want pypitoken to generate the caveats object and pypitoken needs the key and id, the new flow is:

1. we generate a random key and id
2. we generate the pypitoken object
3. we extract the caveats
4. we create the database macaroon with the already computer key, id and caveats

Reply:
The comment above the model field should be updated (it discusses having the default factory in database vs python).

Reply:
(done)